Protection Paladin (Phase 2)

OrcinusDrake opened the pull request:
On feature freeze in preparation for merge

OrcinusDrake committed:
[Paladin] Prot mastery split into 2 ranks

OrcinusDrake committed:
[Paladin] Add holy power to prot

OrcinusDrake committed:
[Paladin] Prot spec auras uneccessary except for the SP conversion

OrcinusDrake committed:
[Paladin] Nothing change. These were just annoying me

OrcinusDrake committed:
[Paladin] Remove Avenger's Valor

OrcinusDrake committed:
[Paladin] Missed some references

OrcinusDrake committed:
[Paladin] First Avenger

OrcinusDrake committed:
[Paladin] Adjusted spell power aura and removed judgment specific
multiplier

OrcinusDrake committed:
[Paladin] Blessed Hammer
Absorbs next damage of any type by value rather than % melee reduction
Also made strikes a float cause int annoyed me.

OrcinusDrake committed:
Merge branch 'shadowlands' into prot_paladin_shadowlands

OrcinusDrake committed:
[Paladin] Redoubt

OrcinusDrake committed:
[Paladin] Melekus' fixes
:embarrassed: Yes I do want sotr to scale with str. Remove redundant 
"gains". Fix BH when strikes < 1 and add upper bound. Redoubt 
simplified. Cache judge hopo. Nuke T20. Misc style+auras

OrcinusDrake committed:
[Paladin] Consecration cd reduction rank

OrcinusDrake committed:
[Paladin] Moment of Glory. Divine Toll separated from AS on charts/stats

OrcinusDrake committed:
[Paladin] Oops, BH strikes were +1 when int Fixed.
Moved judge holy power to constructor. Remove some print tests. Convert 
couple of "find by spell id"s to find by name/variable. Removed 
hardcoded stacks value on redoubt (oops)

OrcinusDrake committed:
[Paladin] SW and naiive implementation of RP.
Have not investigated interaction with holy power reductions.

OrcinusDrake committed:
Merge branch 'shadowlands' into prot_paladin_shadowlands

OrcinusDrake committed:
[Paladin] Judge cd spec aura was duplicated. Add the base 5% dr to cons
Changed some spell ids to find by name

OrcinusDrake committed:
[Paladin] idk how I forgot about goak+RP interaction.
Same naiive implementation as with wings.

OrcinusDrake committed:
[Paladin] Hammer of the righteous 2 charges (from spell rank)

OrcinusDrake committed:
[Paladin] Tiny cons simplification

OrcinusDrake committed:
[Paladin] Try to convert holy_power_consumer to a template. I'm scared

OrcinusDrake committed:
[Paladin] Initialize Word of Glory. Interactions not implemented.

OrcinusDrake committed:
[Paladin] Divine purpose for prot. Cap WoG bonus at 100% missing hp.

OrcinusDrake committed:
[Paladin] Vengeful Shock

OrcinusDrake committed:
[Paladin] Minor fixes hoping to address automatic checks.

OrcinusDrake committed:
Merge branch 'shadowlands' into prot_paladin_shadowlands

OrcinusDrake committed:
[Paladin] wtb Shining Light spelldata

OrcinusDrake committed:
[Paladin] Removed some testing remnants.

OrcinusDrake committed:
Merge branch 'shadowlands' into prot_paladin_shadowlands

OrcinusDrake committed:
[Paladin] Allow composite holy damage on vengeful shock
Remove default spell data nil from holy_power_consumer
Remove some "use_off_gcd"s, restructure BH slightly

OrcinusDrake committed:
[Paladin] Punish the guilty

OrcinusDrake committed:
[Paladin] Bulwark of Righteous Fury
Haven't bothered with range increase.

OrcinusDrake committed:
[Paladin] Holy Avenger's Engraved Sigil

OrcinusDrake committed:
[Paladin] The Ardent Protector's Sanctum

OrcinusDrake committed:
[Paladin] Vanguard's Momentum should be holy damage.

OrcinusDrake committed:
Merge branch 'shadowlands' into prot_pally_SL_phase2
Quick, before it breaks again!

OrcinusDrake committed:
[Paladin] "Badge of the Mad Paragon" renamed to "The Mad Paragon"

Melekus commented on engine/class_modules/paladin/sc_paladin_protection.cpp:
```diff
@@ -113,6 +118,10 @@ struct avengers_shield_t : public avengers_shield_base_t
       p() -> cooldowns.avengers_shield -> reset( false );
       p() -> buffs.moment_of_glory -> decrement( 1 );
     }
+    if ( p() -> legendary.holy_avengers_engraved_sigil -> ok() && rng().roll( holy_avenger_proc_chance ) )
+    {
+      p() -> cooldowns.avengers_shield -> reset( false );
+    }
```
You may want to introduce a proc object similar paladin_t::procs.grand_crusader to track the AS resets from this legendary effect.

Melekus commented on engine/class_modules/paladin/sc_paladin_protection.cpp:
```diff
@@ -439,6 +448,7 @@ void shield_of_the_righteous_buff_t::expire_override( int expiration_stacks, tim
 
 struct shield_of_the_righteous_t : public holy_power_consumer_t<paladin_melee_attack_t>
 {
+  double punish_the_guilty_value;
```
Is it really necessary to use a new value inside sotr instead of just calling p -> conduit.punish_the_guilty.percent() during runtime?

OrcinusDrake replied:
Was in response to https://github.com/simulationcraft/simc/pull/5386#discussion_r501386063 but simplifying it appeals to me. I suppose in comparison to that earlier review the value should already be loaded up in a variable rather than having to look it up.

OrcinusDrake committed:
[Paladin] Count Holy Avenger's Engraved Sigil procs.
Simplify punish the guilty.

OrcinusDrake committed:
[Paladin] Resolute Defender

OrcinusDrake committed:
[Paladin] wtb royal decree spelldata
Do I though?

Melekus commented on engine/class_modules/paladin/sc_paladin_protection.cpp:
```diff
@@ -487,6 +489,14 @@ struct shield_of_the_righteous_t : public holy_power_consumer_t<paladin_melee_at
+    if ( p() -> conduit.resolute_defender -> ok() && p() -> buffs.ardent_defender -> up() )
+    {
+      p() -> buffs.ardent_defender -> extend_duration( p(),
+        p() -> conduit.resolute_defender.percent() * ad_duration
```
You should be able to call p() -> buffs.ardent_defender -> buff_duration() in runtime instead of re-storing the data's duration value in sotr.

OrcinusDrake committed:
[Paladin] Simplify Resolute Defender

OrcinusDrake committed:
[Paladin] Shielding Words

OrcinusDrake committed:
[Paladin] Holy shield uses same formula as normal block.

OrcinusDrake committed:
[Paladin] wtb golden path spelldata

OrcinusDrake committed:
[Paladin] New armour passive

OrcinusDrake committed:
[Paladin] Remove duplicate spec aura hasted gcd/cd implementation.
apply_affecting_aura does that job.

OrcinusDrake committed:
[Paladin] From Dusk till Dawn
It's called "Of Dusk and Dawn" in item_runeforge.inc for some reason. 
It's possible that may change in the future.

OrcinusDrake committed:
[Paladin] Prev Holy Shield fix was wrong. Block value vs block chance

Melekus commented on engine/class_modules/paladin/sc_paladin.cpp:
```diff
@@ -2312,6 +2315,19 @@ double paladin_t::resource_gain( resource_e resource_type, double amount, gain_t
+    if (
+        legendary.from_dusk_till_dawn -> ok() &&
+        amount > 0 &&
+        resources.current[ RESOURCE_HOLY_POWER ] < resources.base[ RESOURCE_HOLY_POWER ] &&
```
It's clearer to use resources.max[] than resources.max[] even if they return the same value.

Melekus replied:
You could also move the whole block below player_t::resource_gain() and only check whether the amount returned by it is != 0 and if you're currently at max.

Melekus commented on engine/class_modules/paladin/sc_paladin.hpp:
```diff
@@ -1104,6 +1112,15 @@ struct holy_power_consumer_t : public Base
 
     ab::execute();
 
+    if (
+        hp_used > 0 &&
+        p -> legendary.from_dusk_till_dawn -> ok() &&
+        p -> resources.current[ RESOURCE_HOLY_POWER ] == p -> legendary.from_dusk_till_dawn -> effectN( 2 ).base_value()
+      )
+    {
+      p -> buffs.blessing_of_dusk -> trigger();
+    }
+
```
Similarly to blessing of dawn, this could be move to paladin_t::resource_loss() after player_t::resource_loss() and would only need to check whether the player is at 0 and if the player actually lost any resource this action.

OrcinusDrake committed:
[Paladin] Prot vanq hammer.
The free sotr currently (incorrectly) activates RP in sim, but that's 
for a future RP fix.

OrcinusDrake committed:
[Paladin] Move both dusk + dawn effects to resource gain/loss

OrcinusDrake committed:
Merge branch 'shadowlands' into prot_pally_SL_phase2

OrcinusDrake committed:
[Paladin] Shining Light
Righteous Protector affected by magistrate's judgment.
Vanq hammer doesn't consume or benefit from divine purpose and has a 
reduced DP proc chance.

OrcinusDrake committed:
[Paladin] Bit more verbose strings on child actions.

OrcinusDrake committed:
Merge branch 'shadowlands' into prot_pally_SL_phase2

OrcinusDrake committed:
[Paladin] Resolute Defender changed. 
Assuming the in-game function is correct rather the tooltip. Assuming 
vanq hammer's auto-sotr procs it but haven't tested it.

OrcinusDrake committed:
[Paladin] Relentless Inquisitor for prot.
Ret you do your own

OrcinusDrake committed:
[Paladin] Shining light doesn't count for RP
Nor any other cost reductions with the exception of divine purpose

OrcinusDrake committed:
[Paladin] The Magistrate's Judgment for prot
Ret is *not* implemented correctly.

OrcinusDrake committed:
[Paladin] Righteous Might Heal

OrcinusDrake committed:
[Paladin] Ashen Hallow
Fixed so it's working now... but it doesn't hit the same haste 
breakpoints.

OrcinusDrake committed:
[Paladin] Put vanq's auto-sotr under vanq hammer (ie. add_child)

OrcinusDrake committed:
[Paladin] Ringing Clarity

OrcinusDrake committed:
[Paladin] Bulwark of Righteous Fury should expire when you use it

OrcinusDrake committed:
[Paladin] Block off time_to_hpg
Could implement in the future, but not interested atm

Melekus commented on engine/class_modules/paladin/sc_paladin.cpp:
```diff
@@ -950,6 +965,13 @@ struct vanquishers_hammer_t : public holy_power_consumer_t<paladin_melee_attack_
     holy_power_consumer_t::impact( s );
 
     p() -> buffs.vanquishers_hammer -> trigger();
+
+    if ( p() -> conduit.righteous_might -> ok() )
+    {
+      r_m_heal -> base_dd_min = r_m_heal -> base_dd_max =
+        s -> result_amount * p() -> conduit.righteous_might.percent();
+      r_m_heal -> execute();
+    }
```
Isn't the conduit value only for the damage increase on VH, with the conversion rate to healing being a static 100%?

OrcinusDrake committed:
[Paladin] Righteous might heals for 100% of damage done.

Melekus commented on engine/class_modules/paladin/sc_paladin.cpp:
```diff
@@ -1003,6 +1002,15 @@ struct divine_toll_t : public paladin_spell_t
+    paladin_spell_t::execute();
+    if ( p() -> conduit.ringing_clarity -> ok() && rng().roll( p() -> conduit.ringing_clarity.percent() ) )
+      for ( int hits = 0; hits < p() -> conduit.ringing_clarity -> effectN( 2 ).base_value(); hits ++ )
+        p() -> active.divine_toll -> schedule_execute();
```
Might want to make sure that this uses the right target, as the tooltip mentions "on the main target".

OrcinusDrake committed:
[Paladin] Display an error and continue gracefully if using time_to_hpg
rather than an assert.

OrcinusDrake committed:
[Paladin] Enforce target for ringing clarity

Melekus commented on engine/class_modules/paladin/sc_paladin.hpp:
```diff
@@ -1145,11 +1145,12 @@ struct holy_power_consumer_t : public Base
-         * num_stacks / 10
-       );
+      // Why is this in deciseconds?
+      timespan_t reduction = -100_ms * p -> talents.righteous_protector -> effectN( 1 ).base_value();
```
Can probabkly just keep using the old format here ``timespan from seconds(value / 10)`` since it's the format that the tooltip code applies. Stupidformattinginterpretation#1 or stupidformattinginterpretation#2 doesn't make much difference anyway.

You should also probably use cost() instead of num_stacks since that one is specific to crusade code (unless you expanded it somewhere else)

OrcinusDrake replied:
Crusade, Righteous Protector, Relentless Inquisitor (the azerite version) (and I suppose fist of justice but NYI) all use the same logic of "only give value for holy power spent", so holy power reductions in general don't proc them. Hence using the same variable

Melekus commented on engine/class_modules/paladin/sc_paladin_protection.cpp:
```diff
@@ -522,6 +537,20 @@ struct shield_of_the_righteous_t : public holy_power_consumer_t<paladin_melee_at
       );
     }
 
+    if ( !background )
```
Could probably simplify this to avoid repeating a lot of similar code.
Maybe something like
```
    {
      if ( p() -> buffs.shining_light_free -> up() || p() -> buffs.shining_light_stacks -> at_max_stacks() )
      {
        p() -> buffs.shining_light_stacks -> expire();
        p() -> buffs.shining_light_free -> trigger();
      }
      else
        p() -> buffs.shining_light_stacks -> trigger();
    }
```

OrcinusDrake committed:
[Paladin] Simplify the Shining Light stack mechanic

OrcinusDrake committed:
[Paladin] vanq hammer's sotr and shining light should not proc RP

Melekus commented on engine/class_modules/paladin/sc_paladin.cpp:
```diff
@@ -1403,7 +1403,7 @@ void paladin_t::init_base_stats()
 
   // Boundless Conviction raises max holy power to 5
-  resources.base[ RESOURCE_HOLY_POWER ] = 3 + passives.boundless_conviction -> effectN( 1 ).base_value();
+  resources.base[ RESOURCE_HOLY_POWER ] = resources.max[ RESOURCE_HOLY_POWER ] = 3 + passives.boundless_conviction -> effectN( 1 ).base_value();
```
resources.max shouldn't need to be set here.

OrcinusDrake replied:
It was set so it could be used in https://github.com/simulationcraft/simc/pull/5411#discussion_r505546437, but after restructuring the flow didn't even need it.

Melekus commented on engine/class_modules/paladin/sc_paladin.cpp:
```diff
@@ -909,27 +933,52 @@ struct hand_of_reckoning_t: public paladin_melee_attack_t
   vanquishers_hammer_t( paladin_t* p, const std::string& options_str ) :
-    holy_power_consumer_t( "vanquishers_hammer", p, p -> covenant.necrolord )
+    holy_power_consumer_t( "vanquishers_hammer", p, p -> covenant.necrolord ),
+    r_m_heal( new righteous_might_t( p ) )
```
Shouldn't need to init r_m_heal if you don't have the conduit.

Melekus commented on engine/class_modules/paladin/sc_paladin.hpp:
```diff
@@ -1126,28 +1127,39 @@ struct holy_power_consumer_t : public Base
-      {
-        num_stacks = 0;
-      }
+        ;
       else
-      {
         p -> buffs.crusade -> trigger( num_stacks );
```
This should probably get reformulated.

OrcinusDrake replied:
Not sure how to make it clearer other than just adding comments. Or replacing `;` with`{ }`. Ret thing I don't really want to touch it

Melekus replied:
You could just reverse the condition so instead of having if (...) {} else x, you'd have if ( ! ... ) { x }

OrcinusDrake replied:
Annoying to read tho. But sure

Melekus commented on engine/class_modules/paladin/sc_paladin_protection.cpp:
```diff
@@ -113,6 +118,11 @@ struct avengers_shield_t : public avengers_shield_base_t
     }
+    if ( p() -> legendary.holy_avengers_engraved_sigil -> ok() && rng().roll( holy_avenger_proc_chance ) )
+    {
+      p() -> cooldowns.avengers_shield -> reset( false );
```
The requires_reaction should be set to true here, since it's only a chance to proc.

OrcinusDrake replied:
Yeah? Even though you have a whole gcd to react and know it's pretty likely?

Melekus replied:
GCD duration is included in the reaction time. 
35% chance to proc is probably high enough to expect the occurence though, you're right.

OrcinusDrake committed:
[Paladin] Only init righteous might when using it

OrcinusDrake committed:
[Paladin] Apply class aura (HoW's cd should be hasted)

OrcinusDrake committed:
[Paladin] These should have different names to select them in the apl

OrcinusDrake committed:
[Paladin] Melekus' suggestion

OrcinusDrake committed:
[Paladin] Don't need to set max hopo

OrcinusDrake committed:
[Paladin] Missed auto-attacks on blessing of dawn

OrcinusDrake committed:
[Paladin] Don't need to set these "gains"

OrcinusDrake committed:
[Paladin] Move bulwark of righteous fury action multiplier
Add comment to Resolute Defender regarding tooltip discrepancy
Remove double that just holds holy avenger's engraved sigil's proc 
chance
Fix Golden Path typo
